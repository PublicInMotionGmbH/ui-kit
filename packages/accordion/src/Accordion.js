import React from 'react'
import PropTypes from 'prop-types'

import { prefix, buildClassName } from '@talixo/shared'

import { Collapse } from '@talixo/collapse'

const propTypes = {
  /** Additional class name */
  className: PropTypes.string,

  /** Should it collapse smoothly? */
  smooth: PropTypes.bool,

  /** Animation time (in ms), requires Collapse geometry CSS */
  animationSpeed: PropTypes.number,

  /** Options to show in accordion */
  options: PropTypes.arrayOf(PropTypes.shape({
    /** Label to show in single option */
    label: PropTypes.node.isRequired,

    /** Content to show inside */
    content: PropTypes.node.isRequired
  })).isRequired,

  /** IDs of currently opened element */
  value: PropTypes.any,

  /** Event fired when button is clicked */
  onChange: PropTypes.func,

  /** Render "open" icon for closed tab */
  renderOpenIcon: PropTypes.func,

  /** Render "close" icon for opened tab */
  renderCloseIcon: PropTypes.func,

  /** Function to build unique ID per option */
  buildId: PropTypes.func,

  /** Should allow opening many sections? */
  multi: PropTypes.bool
}

const defaultProps = {
  animationSpeed: 200,
  smooth: true,
  multi: false,
  buildId: (option, index) => index
}

/**
 * Build element to show in accordion
 *
 * @param {object} props
 * @param {function} props.onOpen
 * @param {function} props.onClose
 * @param {function} props.buildId
 * @param {function} [props.renderOpenIcon]
 * @param {function} [props.renderCloseIcon]
 * @param {array} [props.value]
 * @param {boolean} props.smooth
 * @param {number|null} props.animationSpeed
 * @param {object|{ label: *, content: * }} option
 * @param {number} index
 * @returns {*}
 */
function buildElement (props, option, index) {
  const {
    buildId, onOpen, onClose, value, smooth, animationSpeed,
    renderOpenIcon, renderCloseIcon
  } = props

  // Build ID based on the option
  const id = buildId(option, index)

  // Check if current block is collapsed
  const collapsed = value.indexOf(id) === -1

  // Build onClick handler
  const change = () => collapsed ? onOpen(id) : onClose(id)

  // Build class name for element
  const className = buildClassName('accordion-element', null, { collapsed })

  // Build props for elements
  const buttonProps = {
    role: 'tab',
    className: prefix('accordion-element', 'toggle'),
    onClick: change,
    'aria-expanded': !collapsed
  }

  // Build props for collapse element
  const collapseProps = {
    role: 'tabpanel',
    className: prefix('accordion-element', 'content'),
    smooth: smooth,
    animationSpeed: animationSpeed,
    collapsed: collapsed,
    'aria-hidden': collapsed
  }

  const renderIcon = collapsed ? renderOpenIcon : renderCloseIcon
  const icon = renderIcon ? (
    <span className={prefix('accordion-element', 'toggle-icon')}>
      {renderIcon(option, { ...props, id, collapsed })}
    </span>
  ) : null

  // Build class name for inner elements
  const buttonInnerClsName = prefix('accordion-element', 'toggle-inner')
  const contentInnerClsName = prefix('accordion-element', 'content-inner')

  return (
    <div className={className} key={`${index}--${id}`}>
      <button type='button' {...buttonProps}>
        <div className={buttonInnerClsName}>
          {option.label}
          {icon}
        </div>
      </button>
      <Collapse {...collapseProps}>
        <div className={contentInnerClsName}>
          {option.content}
        </div>
      </Collapse>
    </div>
  )
}

/**
 * Check if both lists contain same elements.
 * This function is intented for user only here,
 * as it has some problems (which do not apply to Accordion),
 * i.e. compare [ 1, 1, 2 ] with [ 1, 2, 2 ]
 *
 * @param {array} prevList
 * @param {array} nextList
 * @returns {boolean}
 */
function isSameList (prevList, nextList) {
  if (prevList.length !== nextList.length) {
    return false
  }

  for (let i = 0; i < prevList.length; i++) {
    if (nextList.indexOf(prevList[i]) === -1) {
      return false
    }
  }

  return true
}

/**
 * Component which represents checkbox.
 *
 * @property {object} props
 * @property {string} [props.className]
 * @property {node} [props.children]
 * @param {function} props.onChange
 * @param {function} props.buildId
 * @param {function} [props.renderOpenIcon]
 * @param {function} [props.renderCloseIcon]
 * @param {*|array} [props.value]
 * @param {boolean} props.smooth
 * @param {boolean} props.multi
 * @param {number} props.animationSpeed
 *
 * @property {object} state
 * @property {array} state.value
 *
 * @class
 */
class Accordion extends React.PureComponent {
  state = {
    value: this.props.value == null ? [] : [].concat(this.props.value)
  }

  /**
   * Update current value in component state,
   * when value is controlled.
   *
   * @param {object} props
   */
  componentWillReceiveProps (props) {
    if (props.value !== undefined && props.value !== this.state.value) {
      this.setState({
        value: props.value == null ? [] : [].concat(props.value)
      })
    }
  }

  /**
   * Change value for self-controlled component,
   * or only send event with new value.
   *
   * @param {*} nextValue
   */
  change = nextValue => {
    const { value, multi, onChange } = this.props
    const { value: prevValue } = this.state

    if (isSameList(prevValue, nextValue)) {
      return
    }

    // Update value when it's self-controlled
    if (value === undefined) {
      this.setState({ value: nextValue })
    }

    if (!multi) {
      nextValue = nextValue.length ? nextValue[0] : null
    }

    // Send event with new value
    if (onChange) {
      onChange(nextValue)
    }
  }

  /**
   * Open section
   *
   * @param {*} id
   */
  onOpen = id => {
    const { multi } = this.props
    const { value } = this.state

    const nextValue = multi ? value.filter(x => x !== id).concat(id) : [ id ]

    this.change(nextValue)
  }

  /**
   * Close section
   *
   * @param {*} id
   */
  onClose = id => {
    const { multi } = this.props
    const { value } = this.state

    const nextValue = multi ? value.filter(x => x !== id) : []

    this.change(nextValue)
  }

  /**
   * Render accordion.
   *
   * @returns {React.Element}
   */
  render () {
    const {
      options, className, smooth, animationSpeed, value, multi,
      buildId, onChange, renderOpenIcon, renderCloseIcon, ...passedProps
    } = this.props

    // Get current value
    const _value = this.state.value

    // Calculate animation time
    const time = parseInt(animationSpeed, 10) || null

    // Build class name for Accordion
    const clsName = buildClassName('accordion', className)

    // Prepare accordion element factory
    const _buildElement = buildElement.bind(null, {
      onOpen: this.onOpen,
      onClose: this.onClose,
      buildId: buildId,
      value: _value,
      smooth: smooth,
      animationSpeed: time,
      renderOpenIcon: renderOpenIcon,
      renderCloseIcon: renderCloseIcon
    })

    // Build accordion elements
    const elements = options.map(_buildElement)

    return (
      <div className={clsName} {...passedProps}>
        {elements}
      </div>
    )
  }
}

Accordion.displayName = 'Accordion'

Accordion.propTypes = propTypes
Accordion.defaultProps = defaultProps

export default Accordion
