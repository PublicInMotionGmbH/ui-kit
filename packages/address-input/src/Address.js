import React from 'react'
import PropTypes from 'prop-types'

import { Icon } from '@talixo/icon'
import { buildClassName } from '@talixo/shared'

export const moduleName = 'address'

const defaultProvider = {
  bar: 'local_bar',
  cafe: 'local_cafe',
  car_wash: 'local_car_wash',
  food: 'local_dining',
  restaurant: 'local_dining',
  drink: 'local_drink',
  gas_station: 'local_gas_station',
  hotel: 'local_hotel',
  hospital: 'local_hospital',
  store: 'local_grocery_store',
  activity: 'local_play',
  airport: 'local_airport',
  parking: 'local_parking',
  central_train_station: 'train',
  poi_train_station: 'train',
  transit_station: 'train',
  subway_station: 'train',
  train_station: 'train',
  light_rail_station: 'train',
  bus_station: 'directions_bus',
  taxi_stand: 'local_taxi',
  atm: 'local_atm',
  pharmacy: 'local_pharmacy',
  default: 'location_on'
}

const propTypes = {

  /** Address of a place. */
  address: PropTypes.string,

  /** Additional class name passed to wrapper. */
  className: PropTypes.string,

  /** Place details. */
  details: PropTypes.string,

  /** Format details. */
  detailsFormatter: PropTypes.func,

  /** Icon displayed next to the address. */
  icon: PropTypes.string,

  /** You can provide here your custom mapping of Icon component names. Remember to add fallback icon under `default` key. */
  iconProvider: PropTypes.object,

  /** Place name abbreviation. It can be e.g. IATA code of an airport. */
  short: PropTypes.string
}

const defaultProps = {
  detailsFormatter: (x) => (x),
  iconProvider: defaultProvider
}

/**
 * This is a component which represents a place Address.
 *
 * @param props
 * @param {string} [props.abbreviation]
 * @param {string} [props.address]
 * @param {string} [props.className]
 * @param {string} [props.details]
 * @param {string} [props.type]
 *
 * @returns {React.Element}
 */
function Address (props) {
  const { address, className, details, detailsFormatter, icon, iconProvider, meta, short, ...passedProps } = props

  // Get classNames
  const shortCls = buildClassName([moduleName, 'short'])
  const detailsCls = buildClassName([moduleName, 'details'])
  const iconCls = buildClassName([moduleName, 'icon'])
  const wrapperCls = buildClassName(moduleName, className)

  // Get Icon element
  const iconType = iconProvider[icon] || iconProvider.default
  const iconElement = <Icon name={iconType} />
  const detailsElement = detailsFormatter(details)
  const abbreviationElement = short == null ? null : <span className={shortCls}>{short.toUpperCase()}</span>

  return (
    <div className={wrapperCls} {...passedProps}>
      <span className={iconCls}>{iconElement}</span>
      {abbreviationElement}
      <span className={detailsCls}>
        <strong>{address}</strong>
        <span>{detailsElement}</span>
      </span>
    </div>
  )
}

Address.propTypes = propTypes
Address.defaultProps = defaultProps

export default Address
