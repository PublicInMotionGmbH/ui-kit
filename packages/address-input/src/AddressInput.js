import React from 'react'
import PropTypes from 'prop-types'
import { findDOMNode } from 'react-dom'

import debounce from 'lodash/debounce'
import { TransitionGroup, CSSTransition } from 'react-transition-group'

import { AutoComplete } from '@talixo/combo-box'
import { TextInput } from '@talixo/text-input'
import { ProgressRing } from '@talixo/progress-ring'
import { MaskedInput } from '@talixo/masked-input'
import { DeviceSwap } from '@talixo/device-swap'
import { Button } from '@talixo/button'
import { Icon } from '@talixo/icon'
import { buildClassName } from '@talixo/shared'

import { Address } from '@talixo/address'

export const moduleName = 'address-input'

/**
 * Get description either from meta data or from address property of value.
 *
 * @param {object} value
 * @returns {string}
 */
export function getDescription (value) {
  if (value) {
    if (value.meta) {
      return value.meta.description
    } else if (value.address) {
      return value.address
    }
  }

  return ''
}

/**
 * Get string from an object. To use with AutoComplete component.
 *
 * @param {object} item
 *
 * @returns {string}
 */
export function itemToString (item) {
  return item == null ? '' : getDescription(item)
}

/**
 *
 * @param {object} props
 * @param {string} [props.address]
 * @param {string} [props.className]
 * @param {string} [props.details]
 * @param {function} [props.formatDetails]
 * @param {string} [props.type]
 * @param {string} [props.short]
 *
 * @returns {React.Element}
 */
function _renderAddress (props) {
  const { address, details, formatDetails, type, short } = props

  return React.createElement(Address, {
    address, details, formatDetails, type, short
  })
}

const locationProp = {
  /** Address of location. */
  address: PropTypes.string,

  /** Additional information about place. */
  details: PropTypes.string,

  /** Icon displayed next to the address. */
  type: PropTypes.string,

  /** Metadata of location. */
  meta: PropTypes.shape({
    /** Description of a place. If provided it will be displayed inside AddressInput. */
    description: PropTypes.string
  }),

  /** Place name abbreviation. It can be e.g. IATA code of an airport. */
  short: PropTypes.string
}

const propTypes = {
  /** Additional className passed to wrapper. */
  className: PropTypes.string,

  /** AutoComplete list footer. */
  footer: PropTypes.node,

  /** Mobile view input label. */
  label: PropTypes.string,

  /** Indicates if loader should be displayed inside input. */
  loading: PropTypes.bool,

  /** Locations to be displayed inside AutoComlete. */
  locations: PropTypes.arrayOf(PropTypes.shape(locationProp)),

  /** onBlur callback. */
  onBlur: PropTypes.func,

  /** onChange callback. */
  onChange: PropTypes.func,

  /** onFocus callback. */
  onFocus: PropTypes.func,

  /**
   * This function is called when changes input value, at has typed in at least 3 (minLetters) letters.
   * It should be used to load locations from external API.
   */
  onLoadRequest: PropTypes.func,

  /**
   * This function is called when changes input value, at has typed less than 3 (minLetters) letters.
   * It should be used to prevent request to API.
   */
  onStopRequest: PropTypes.func,

  /** Minimum number of letters to run load/stop helpers */
  minLetters: PropTypes.number,

  /** Delay time for writing message */
  writingDelay: PropTypes.number,

  /** AddressInput placeholder. */
  placeholder: PropTypes.string,

  /** Address component which will be displayed inside autocomplete list. */
  renderAddress: PropTypes.func,

  /** Chosen location value. */
  value: PropTypes.shape(locationProp),

  /** Should it be optimized for mobile? */
  mobileFriendly: PropTypes.bool
}

const defaultProps = {
  renderAddress: _renderAddress,
  mobileFriendly: true,
  minLetters: 3,
  writingDelay: 300
}

/**
 * Component which represents AddressInput.
 *
 * @property {object} props
 * @property {string} [props.className]
 * @property {*} [props.footer]
 * @property {string} [props.label]
 * @property {boolean} [props.loading]
 * @property {Array<{ [address]: string, [details]: string, [type]: string, [meta]: { [description]: string }, [short]: string}>} [props.locations]
 * @property {function} [props.onBlur]
 * @property {function} [props.onChange]
 * @property {function} [props.onFocus]
 * @property {function} [props.onLoadRequest]
 * @property {function} [props.onStopRequest]
 * @property {function} [props.placeholder]
 * @property {function} [props.renderAddress]
 *
 * @property {object} [props.value]
 * @property {string} [props.value.address]
 * @property {string} [props.value.details]
 * @property {string} [props.value.icon]
 * @property {object} [props.value.meta]
 * @property {string} [props.value.meta.description]
 * @property {string} [props.value.short]
 *
 * @property {string} state
 * @property {string} state.inputValue
 * @property {object} state.value
 *
 * @class
 */
class AddressInput extends React.PureComponent {
  state = {
    inputValue: getDescription(this.props.value),
    value: this.props.value || null
  }

  componentWillReceiveProps (props) {
    if (props.value !== this.props.value && props.value !== undefined) {
      const state = { value: props.value }

      if (props.value) {
        state.inputValue = getDescription(props.value)
      }

      this.setState(state)
    }
  }

  /**
   * Handle user typing.
   *
   * @param {string} value
   */
  onInputValueChange = value => {
    const { onLoadRequest, onStopRequest, minLetters } = this.props

    if (value === this.state.inputValue) {
      return
    }

    this.setState({ inputValue: value })
    this.onChange(null)

    // Invoke find request only when at least 3 chars are provided.
    if (value.length < minLetters) {
      this.load.cancel()

      if (onStopRequest) {
        onStopRequest()
      }
    } else if (onLoadRequest) {
      this.load(value)
    }
  }

  /**
   *
   * @type {Function}
   */
  load = debounce(value => {
    this.props.onLoadRequest(value)
  }, this.props.writingDelay)

  /**
   * Handle choosing element from the autocomplete list on mobile devices.
   *
   * @param {object} value
   */
  onChoose = value => {
    this.setState({ focus: false })
    this.onChange(value)
    this.input.blur()
  }

  /**
   * Handle choosing element from autocomplete list.
   * @param value
   */
  onChange = value => {
    if (this.props.value === undefined) {
      this.setState({ value })
    }

    if (this.props.onChange) {
      this.props.onChange(value)
    }
  }

  /**
   * Handle input focusing.
   */
  focus = () => {
    this.setState({
      focus: true
    })

    if (this.props.onFocus) {
      this.props.onFocus()
    }
  }

  /**
   * Handle input blur event.
   */
  blur = () => {
    const state = { focus: false }
    const { locations, onChange, onBlur } = this.props

    // If user types valid location set it as a value
    if (!this.state.value && locations && locations.length) {
      const firstValue = locations[0]

      state.value = firstValue
      state.inputValue = getDescription(firstValue)

      if (onChange) {
        onChange(firstValue)
      }
    }

    // Set proper search query to input.
    if (this.state.value) {
      state.inputValue = getDescription(this.state.value)
    }

    this.setState(state)

    if (onBlur) {
      onBlur()
    }
  }

  /**
   * Clear input and location search. Focus current input if possible.
   */
  onInputClear = () => {
    this.onInputValueChange('')

    if (this.input) {
      this.input.focus()
    }
  }

  /**
   * Handle tab pressing. It will set first location as value if locations array is provided.
   * @param {Event|SyntheticEvent} event
   */
  onKeyDown = (event) => {
    const { locations } = this.props

    // 'Tab' key
    if (event.which === 9 && locations && locations.length === 1) {
      this.onChange(locations[0])
    }
  }

  /**
   * Get reference to mobile version of AddressInput
   *
   * @param {HTMLElement} node
   */
  setInputRef = (node) => {
    this.input = findDOMNode(node)
  }

  /**
   * Render desktop version of AddressInput.
   *
   * @returns {React.Element}
   */
  renderDesktop = () => {
    const {
      className, footer, locations, loading, onBlur, onChange, onFocus, writingDelay,
      onLoadRequest, onStopRequest, placeholder, renderAddress, value, minLetters, mobileFriendly,
      ...passedProps
    } = this.props

    // FIXME: pass props down
    return (
      <MaskedInput
        value={this.state.value}
        renderMask={() => renderAddress(this.state.value)}
        {...passedProps}
      >
        <AutoComplete
          openOnFocus
          options={locations}
          onChoose={this.onChoose}
          onFocus={this.focus}
          onBlur={this.blur}
          renderItem={renderAddress}
          inputValue={this.state.inputValue}
          onInputValueChange={this.onInputValueChange}
          itemToString={itemToString}
          footer={footer}
        >
          <TextInput
            inputRef={this.setInputRef}
            autoComplete='nope'
            spellCheck='false'
            onKeyDown={this.onKeyDown}
            placeholder={placeholder}
            right={loading ? <ProgressRing /> : null}
          />
        </AutoComplete>
      </MaskedInput>
    )
  }

  /**
   * Render mobile version of AddressInput.
   *
   * @returns {*}
   */
  renderMobile = () => {
    const { footer, label, locations, loading, placeholder, renderAddress } = this.props
    const fakeCls = buildClassName([moduleName, 'fake-input'])
    const transitionCls = buildClassName([moduleName, 'transition'])
    const mobileModal = buildClassName([moduleName, 'mobile-modal'])
    const mobileModalHeader = buildClassName([moduleName, 'mobile-modal-header'])

    const renderAddressProps = { ...this.state.value }

    const input = this.state.focus ? (
      <CSSTransition
        classNames={transitionCls}
        timeout={200}
      >
        <div className={mobileModal}>
          <div className={mobileModalHeader}>
            <div>{label}</div>
            <Button small onClick={this.blur}>Close</Button>
          </div>
          <AutoComplete
            openOnFocus
            isOpen
            options={locations}
            onChoose={this.onChoose}
            onFocus={this.focus}
            renderItem={renderAddress}
            inputValue={this.state.inputValue}
            onInputValueChange={this.onInputValueChange}
            itemToString={itemToString}
            footer={footer}
          >
            <TextInput
              autoComplete='nope'
              spellCheck='false'
              onKeyDown={this.onKeyDown}
              placeholder={placeholder}
              right={loading ? <ProgressRing /> : <Icon name='clear' onClick={this.onInputClear} />}
              inputRef={this.setInputRef}
              autoFocus
            />
          </AutoComplete>
        </div>
      </CSSTransition>
    ) : null

    return (
      <div>
        <div className={fakeCls} onClick={this.focus}>
          {this.state.value ? renderAddress(renderAddressProps) : placeholder}
        </div>
        <TransitionGroup>
          {input}
        </TransitionGroup>
      </div>
    )
  }

  render () {
    const { className, mobileFriendly } = this.props

    const clsName = buildClassName(moduleName, className, {
      focused: this.state.focus
    })

    const input = mobileFriendly ? (
      <DeviceSwap
        renderDesktop={this.renderDesktop}
        renderMobile={this.renderMobile}
      />
    ) : this.renderDesktop()

    return (
      <div className={clsName}>
        {input}
      </div>
    )
  }
}

AddressInput.displayName = 'AddressInput'

AddressInput.propTypes = propTypes
AddressInput.defaultProps = defaultProps

export default AddressInput
