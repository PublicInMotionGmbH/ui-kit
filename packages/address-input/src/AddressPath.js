import React from 'react'
import PropTypes from 'prop-types'

import { buildClassName } from '@talixo/shared'

const moduleName = 'address-path'

const propTypes = {
  /** Address Elements to be displayed inside path. */
  children: PropTypes.node.isRequired,

  /** Additional class name passed to wrapper. */
  className: PropTypes.string
}

/**
 *
 * @param props
 * @param {*} children
 * @param {string} className
 * @returns {React.Element}
 */
function AddressPath (props) {
  const { children, className, ...passedProps } = props

  const lastIndex = children.length - 1
  const lastPathIndex = lastIndex - 1

  // Generate classNames
  const wrapper = buildClassName(moduleName, className)
  const pathCls = buildClassName([moduleName, 'path'])
  const elementCls = index => buildClassName([moduleName, 'point'], null, { 'last-path': lastPathIndex === index })

  // Generate Address elements and apply path styles
  const elements = React.Children.map(children, (element, index) => (
    <div className={elementCls(index)}>
      {index === lastIndex ? null : <span className={pathCls} />}
      {element}
    </div>
  ))

  return (
    <div className={wrapper} {...passedProps}>
      {elements}
    </div>
  )
}

AddressPath.propTypes = propTypes

export default AddressPath
