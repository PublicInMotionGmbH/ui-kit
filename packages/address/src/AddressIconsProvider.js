import React from 'react'
import PropTypes from 'prop-types'

import defaultIconsMap from './defaultIconsMap'

export const moduleName = 'address'

const propTypes = {
  /** Map of address type => corresponding icon name */
  types: PropTypes.objectOf(PropTypes.string),

  /** Components which should be rendered with this context. */
  children: PropTypes.node
}

const defaultProps = {
  types: defaultIconsMap
}

const childContextTypes = {
  addressTypesIconMap: PropTypes.objectOf(PropTypes.string)
}

/**
 * This is a component which represents a place Address.
 *
 * @param {object} props
 * @param {object} [props.types]
 * @param {*} props.children
 *
 * @returns {React.Element}
 */
class AddressIconsProvider extends React.Component {
  getChildContext () {
    return {
      addressTypesIconMap: this.props.types
    }
  }

  render () {
    return this.props.children
  }
}

AddressIconsProvider.displayName = 'AddressIconProvider'

AddressIconsProvider.childContextTypes = childContextTypes
AddressIconsProvider.propTypes = propTypes
AddressIconsProvider.defaultProps = defaultProps

export default AddressIconsProvider
