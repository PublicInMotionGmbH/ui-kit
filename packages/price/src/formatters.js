/**
 * Converts given value to float value with fixed precision.
 *
 * @param {string|number} value
 * @param {number} precision
 * @returns {string}
 */
export function valueToFixedPrecision (value, precision = 2) {
  return parseFloat(value).toFixed(precision)
}

/**
 * It tries to format number using JS Intl and returns formatted number.
 *
 * @param {string|number} value
 * @param {string} locale
 * @param {number} precision
 * @returns {string}
 */
export function formatNumber (value, locale, precision) {
  try {
    const price = new Intl.NumberFormat(locale, {
      minimumFractionDigits: precision,
      maximumFractionDigits: precision }).format(value)
    return price
  } catch (e) { return value }
}

/**
 * Formats price according to provided mapping or locale and currency
 *
 * @param {string|number} value
 * @param {object} config
 * @param {string} [config.currency]
 * @param {object} [config.currencyToSymbol]
 * @param {number} [config.precision]
 * @param {string} [config.locale]
 * @returns {object}
 */
export function generatePrice (value, config = {}) {
  const { currency, currencyToSymbol, locale, precision } = config

  value = valueToFixedPrecision(value, precision)

  if (isNaN(value)) {
    return {
      value: value,
      symbol: currency,
      error: true
    }
  }

  // If user provides own mapping, use it.
  if (currencyToSymbol) {
    const symbol = currencyToSymbol[currency] || currency
    const price = formatNumber(value, locale, precision)

    return { value: price, symbol }
  } else if (locale && currency) {
    // If user provides locale and currency, try to format it according to provided information using JS Intl
    try {
      const formattedValue = new Intl.NumberFormat(locale, {
        style: 'currency',
        currency: currency,
        minimumFractionDigits: precision,
        maximumFractionDigits: precision
      }).format(value)

      // Try to extract symbol and value
      const matchedValue = formattedValue.match(/^(.*?)((?:[0-9.,-]|\s)+)(.*)$/)

      if (matchedValue) {
        const symbolBefore = matchedValue[1].trim()
        const value = matchedValue[2].trim()
        const symbolAfter = matchedValue[3].trim()

        if (value && (symbolAfter || symbolBefore)) {
          return {
            value: value,
            symbol: symbolAfter || symbolBefore,
            displayBefore: !!symbolBefore
          }
        }
      }

      return { value: formattedValue, symbol: null }
    } catch (e) {
      // Fallback if js Intl is not supported by a browser
      const price = formatNumber(value, locale, precision)
      return { value: price, symbol: currency }
    }
  } else {
    const price = formatNumber(value, locale, precision)
    return { value: price, symbol: currency }
  }
}
