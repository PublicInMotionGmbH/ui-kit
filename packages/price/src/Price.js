import React from 'react'
import PropTypes from 'prop-types'

import { buildClassName } from '@talixo/shared'

import { generatePrice } from './formatters'

const moduleName = 'price'

const propTypes = {
  /** Additional class name passed to wrapper. */
  className: PropTypes.string,

  /** Currency of a price. */
  currency: PropTypes.string,

  /** Mapping which provides will be used to convert currency to provided symbol. */
  currencyToSymbol: PropTypes.object,

  /** Should the currency be displayed before price? */
  displayBefore: PropTypes.bool,

  /** This placeholder is displayed if provided value is not a number (isNaN returns true). */
  errorPlaceholder: PropTypes.node,

  /** Locale code of user. */
  locale: PropTypes.string,

  /** Price precision. */
  precision: PropTypes.number,

  /** Prefix whoich will be displayed before the price. Can be to e.g. indicate that price is aproximate. */
  prefix: PropTypes.node,

  /** Given price. */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}

const defaultProps = {
  displayBefore: false,
  errorPlaceholder: '-',
  precision: 2
}

/**
 * Component which represents Price.
 *
 * @param {object} props
 * @param {string} [props.className]
 * @param {string} [props.currency]
 * @param {object} [props.currencyToSymbol]
 * @param {boolean} [props.displayBefore]
 * @param {*} [props.errorPlaceholder]
 * @param {string} [props.locale]
 * @param {number} [props.precision]
 * @param {*} [props.prefix]
 * @param {string|number} [props.value]
 *
 * @returns {React.Element}
 */
function Price (props) {
  const {
    className, currency, currencyToSymbol, decimal, errorPlaceholder,
    displayBefore, locale, prefix, precision, value, ...passedProps
  } = props
  const wrapperCls = buildClassName(moduleName, className)
  const prefixElem = prefix ? <small>{prefix}</small> : null

  // Get formatted price and symbol if needed.
  const price = generatePrice(value, {
    locale, currency, currencyToSymbol, precision
  })

  // Generate price element
  const priceElement = price.error
    ? <span>{errorPlaceholder}</span>
    : <span>
      {displayBefore && price.symbol}
      {price.value}
      {!displayBefore && price.symbol}
    </span>

  return (
    <span className={wrapperCls} {...passedProps}>
      {prefixElem}
      {priceElement}
    </span>
  )
}

Price.displayName = 'Price'

Price.propTypes = propTypes
Price.defaultProps = defaultProps

export default Price
