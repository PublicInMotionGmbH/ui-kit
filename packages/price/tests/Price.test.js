import React from 'react'
import { shallow } from 'enzyme'

import Price from '../src/Price'
import { formatNumber, generatePrice, valueToFixedPrecision } from '../src/formatters'

// Helpers
const currencyToSymbol = {
  USD: '$',
  EUR: 'â‚¬'
}

// Tests
describe('<Price />', () => {
  describe('<Price /> comoponent', () => {
    it('should render children correctly', () => {
      const wrapper = shallow(<Price currency='PLN' value={1000} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should render price with curency before price', () => {
      const wrapper = shallow(<Price currency='GBP' value={1000} displayBefore />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should render prefix correctly', () => {
      const wrapper = shallow(<Price currency='PLN' value={1000} prefix='about' />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should render errorPlaceholder when wrong value is provided', () => {
      const wrapper = shallow(<Price currency='PLN' value='asd' />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('formatters', () => {
    describe('value to fixed precision', () => {
      it('should return "1.22" given 1.2222222', () => {
        expect(valueToFixedPrecision(1.2222222)).toBe('1.22')
      })

      it('should return "1.22" given "1.2222222"', () => {
        expect(valueToFixedPrecision('1.2222222')).toBe('1.22')
      })

      it('should return 2 given 1.500000', () => {
        expect(valueToFixedPrecision('1.500000', 0)).toBe('2')
      })

      it('should return 1.6 given 1.55', () => {
        expect(valueToFixedPrecision('1.55', 1)).toBe('1.6')
      })

      it('should return NaN when value is not a number', () => {
        expect(valueToFixedPrecision('adsadasd', 4)).toBe('NaN')
      })
    })

    describe('fomatNumber', () => {
      it('should return 1.222 given 1.2222321213', () => {
        expect(formatNumber('1.2222321213', 'en', 3)).toBe('1.222')
      })

      it('should return 1,000,000 given 1000000000.2121', () => {
        expect(formatNumber('1000000000.2121', 'en', 0)).toBe('1,000,000,000')
      })

      it('should return 1,000,000 given 1000000000.2121', () => {
        expect(formatNumber('1000000000.2121', 'en', 3)).toBe('1,000,000,000.212')
      })

      it('should return 1000000000.2121 given 1000000000.2121 when locale is null', () => {
        expect(formatNumber('1000000000.2121', null, 3)).toBe('1000000000.2121')
      })

      it('should return 1000000000.2121 given 1000000000 when precision is null', () => {
        expect(formatNumber('1000000000.2121', 'en', null)).toBe('1,000,000,000')
      })
    })

    describe('generatePrice', () => {
      it('should set error to true when provided value isNaN', () => {
        expect(generatePrice('asddsa').error).toBe(true)
      })

      it('should return mapped currency symbol and formatted value', () => {
        const result = generatePrice('100000', { currencyToSymbol, currency: 'USD' })
        expect(result.value).toBe('100,000')
        expect(result.symbol).toBe(currencyToSymbol['USD'])
        expect(result.error).toBe(undefined)
      })

      it('should return currency string when its not mapped and formatted value', () => {
        const result = generatePrice('100000', { currencyToSymbol, currency: 'PLN' })
        expect(result.value).toBe('100,000')
        expect(result.symbol).toBe('PLN')
        expect(result.error).toBe(undefined)
      })

      it('should return $100,000.12 given locale, value and currency', () => {
        const result = generatePrice('100000.12', { locale: 'en', currency: 'USD' })
        expect(result.value).toBe('100,000.12')
        expect(result.symbol).toBe('$')
        expect(result.displayBefore).toBe(true)
        expect(result.error).toBe(undefined)
      })

      it('should return 100,000.12 and USD given value and currency and locale but when error occurs', () => {
        const _backup = Intl.NumberFormat
        Intl.NumberFormat = undefined
        const result = generatePrice('100000.12', { locale: 'en', currency: 'USD' })
        expect(result.value).toBe('100000.12')
        expect(result.symbol).toBe('USD')
        expect(result.error).toBe(undefined)
        Intl.NumberFormat = _backup
      })
    })
  })
})
