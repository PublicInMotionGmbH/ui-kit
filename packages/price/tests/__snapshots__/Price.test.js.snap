// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Price /> <Price /> comoponent should render children correctly 1`] = `
<span
  className="talixo-price"
>
   
  1,000.00
   
  <abbr
    title="PLN"
  >
    PLN
  </abbr>
</span>
`;

exports[`<Price /> <Price /> comoponent should render errorPlaceholder when wrong value is provided 1`] = `
<span
  className="talixo-price"
>
  -
</span>
`;

exports[`<Price /> <Price /> comoponent should render prefix correctly 1`] = `
<span
  className="talixo-price"
>
  <small>
    about
  </small>
   
  1,000.00
   
  <abbr
    title="PLN"
  >
    PLN
  </abbr>
</span>
`;

exports[`<Price /> <Price /> comoponent should render price with curency before price 1`] = `
<span
  className="talixo-price"
>
  <abbr
    title="GBP"
  >
    GBP
  </abbr>
   
  1,000.00
   
</span>
`;
